test/functional/testlib.bash

```diff
@@ -0,0 +1,502 @@
+#!bin/bash
+
+FUNC_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+TEST_FILENAME=$(basename "$BATS_TEST_FILENAME")
+TEST_NAME=${TEST_FILENAME%.bats}
+export TEST_NAME
+export FUNC_DIR
+export SWUPD_DIR="$FUNC_DIR/../.."
+export SWUPD="$SWUPD_DIR/swupd"
+
+
+generate_random_content() { 
+
+	local bottom_range=5
+	local top_range=100
+	local range=$((top_range - bottom_range + 1))
+	local number_of_lines=$((RANDOM%range))  
+	< /dev/urandom tr -dc 'a-zA-Z0-9-_!@#$%^&*()_+{}|:<>?=' | fold -w 100 | head -n $number_of_lines
+
+}
+
+generate_random_name() { 
+
+	local prefix=${1:-test-}
+	local uuid
+	
+	# generate random 8 character alphanumeric string (lowercase only)
+	uuid=$(< /dev/urandom tr -dc 'a-f0-9' | fold -w 8 | head -n 1)
+	echo "$prefix$uuid"
+
+}
+
+validate_path() { 
+
+	local path=$1
+	if [ -z "$path" ] || [ ! -d "$path" ]; then
+		echo "Please provide a valid path"
+		return 1
+	fi
+
+}
+
+validate_item() { 
+
+	local vfile=$1
+	if [ -z "$vfile" ] || [ ! -e "$vfile" ]; then
+		echo "Please provide a valid file"
+		return 1
+	fi
+
+}
+
+validate_param() {
+
+	local param=$1
+	if [ -z "$param" ]; then
+		echo "Mandatory parameter missing"
+		return 1
+	fi
+
+}
+
+# Exports environment variables that are dependent on the test environment
+# Parameters:
+# - ENV_NAME: the name of the test environment
+set_env_variables() {
+
+	local env_name=$1
+	validate_path "$env_name"
+
+	export SWUPD_OPTS="-S$BATS_TEST_DIRNAME/$env_name/state -p $BATS_TEST_DIRNAME/$env_name/target-dir -F staging -u file://$BATS_TEST_DIRNAME/$env_name/web-dir -C $FUNC_DIR/Swupd_Root.pem -I"
+
+}
+
+# Creates a directory with a hashed name in the specified path, if a directory
+# already exists it returns the name
+# Parameters:
+# - PATH: the path where the directory will be created 
+create_dir() { 
+	
+	local path=$1
+	local hashed_name
+	local directory
+	validate_path "$path"
+	
+	# most directories have the same hash, so we only need one directory
+	# in the files directory, if there is already one just return the path/name
+	directory=$(find "$path"/* -type d 2> /dev/null)
```

Comment:
This worries me. What happens if we have two directories or more directories under $path?

Reply:
it should not happen, in general the content of the whole test environment is created through the test library, so we can control what goes where, and test environments are created and destroyed with every test....  and even if it happened because someone did this in one test (which would be a bad practice that could be caught when reviewing tests that are submitted) that test won't work properly, so the author of the test would only need to fix its test.

Reply:
the $path that will be sent to this function will always be tests/functional/<test category>/<test theme>/<test environment>/web-dir/<version>/files, and this directory will always have only one directory